subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java

```diff
@@ -1053,18 +1053,15 @@ public final ExecutionResult run() {
             if (errorsShouldAppearOnStdout()) {
                 result = new ErrorsOnStdoutScrapingExecutionResult(result);
             }
+            afterExecute.execute(this);
```

Comment:
This changes the contract of `afterExecute` though I am not sure if that contract was explicit or even relied upon.

Should we not leverage _suppressed exceptions_ instead? Assuming we can use Java 7 features in this code.

Reply:
I do not think changing the contract is a problem. If an exception is thrown in the first block it means already something is wrong and the test should fail with error. And any follow-up exception/error in the second block might be a _false negative_ as it is the case in the example I investigated. I also think it is not necessary to execute the code in the _afterExecute_ if the test already failed before. It should not contain any logic that is necessary to run when the test failed (in the implementations I checked, we do additional assertions or collect data for assertions in the test).

Anyway, my main concern was making the code not unnecessary complicated. I wasn't aware _suppressed exceptions_ can be used here. How would it look like in this case?

Reply:
Two options:

1. Introduce a `Closeable` which becomes a resource of the `try` block and invokes `afterExecute` in its `close` method.
2. Do it by hand, tracking any exception thrown, invoking `afterExecute` and collect exceptions from it into any original thrown exception or rethrow if no prior exception.

But really the first one should be used for simplicity :-)
Though given your analysis, it might not be necessary at all.

Reply:
(2.) is basically what I also thought about. And yes (1) would be simpler to understand. Good to know, thanks. Still it would make it more complicated (it's not a one liner).

If you are fine to leave the change as it is, please approve the PR. Then I can merge it to `master`.